### PR TITLE
OZ L02; Spearbit 87; improved calldata checks

### PR DESCRIPTION
**Pull request by hensha256**

### CODE REVIEW

File: src/libraries/CalldataDecoder.sol [hunk 1]
```diff
@@ -14,34 +14,18 @@ library CalldataDecoder {
     /// @notice equivalent to SliceOutOfBounds.selector
     bytes4 constant SLICE_ERROR_SELECTOR = 0x3b99b53d;
```
Comment (HickupHH3) [inline src/libraries/CalldataDecoder.sol:15]: redundant

File: src/libraries/CalldataDecoder.sol [hunk 2]
```diff
@@ -237,38 +221,69 @@ library CalldataDecoder {
...
     /// @param _arg The index of the argument to extract
-    function toBytes(bytes calldata _bytes, uint256 _arg) internal pure returns (bytes calldata res) {
-        (uint256 length, uint256 offset) = toLengthOffset(_bytes, _arg);
+    function toBytesArray(bytes calldata _bytes, uint256 _arg) internal pure returns (bytes[] calldata res) {
+        uint256 encodingLength;
+        bool isValidEncoding = true;
         assembly ("memory-safe") {
-            res.length := length
-            res.offset := offset
+            let bytesOffset := and(_bytes.offset, OFFSET_OR_LENGTH_MASK)
```
Comment (kmbarry1) [inline src/libraries/CalldataDecoder.sol:254]: Masking done here but not in `toBytes`--intentional?

Comment (hensha256) [reply src/libraries/CalldataDecoder.sol:252]: have pushed an update for that https://github.com/Uniswap/v4-periphery/pull/316/commits/2fd239fc0b585a3809e088f78c82ded827abe80e

**Review by 0age — DISMISSED** [review]

File: src/libraries/CalldataDecoder.sol [hunk 3]
```diff
@@ -237,38 +221,69 @@ library CalldataDecoder {
...
-        pure
-        returns (uint256 length, uint256 offset)
-    {
+    function toBytes(bytes calldata _bytes, uint256 _arg) internal pure returns (bytes calldata res) {
+        uint256 length;
         uint256 relativeOffset;
         assembly ("memory-safe") {
             // The offset of the `_arg`-th element is `32 * arg`, which stores the offset of the length pointer.
             // shl(5, x) is equivalent to mul(32, x)
             let lengthPtr := add(_bytes.offset, calldataload(add(_bytes.offset, shl(5, _arg))))
```
[the remark this answers is not in the document]
Comment (hensha256) [reply src/libraries/CalldataDecoder.sol]: i think youre commenting on an old version - i apply the mask now. Added in https://github.com/Uniswap/v4-periphery/commit/2fd239fc0b585a3809e088f78c82ded827abe80e

[the remark this answers is not in the document]
Comment (hensha256) [reply src/libraries/CalldataDecoder.sol]: oh do you mean mask `arg`? i'm not too concerned about that as we will always pass in that in internally to the code

[the remark this answers is not in the document]
Comment (0age) [reply src/libraries/CalldataDecoder.sol]: yeah i think this comment is outdated now!

**Review by 0age — DISMISSED** [review]

File: test/libraries/CalldataDecoder.t.sol [hunk 4]
```diff
@@ -152,6 +153,52 @@ contract CalldataDecoderTest is Test {
...
+        _actionParams[0] = hex"11111111";
+        _actionParams[1] = hex"22";
+        _actionParams[2] = hex"3333333333333333";
+        _actionParams[3] = hex"4444444444444444444444444444444444444444444444444444444444444444";
+
+        bytes memory params = abi.encode(_actions, _actionParams);
+
+        bytes memory invalidParams = new bytes(params.length - 1);
+        // dont copy the final byte
+        for (uint256 i = 0; i < params.length - 2; i++) {
```
Comment (snreynolds) [inline test/libraries/CalldataDecoder.t.sol:179]: ayyyyy

File: src/libraries/CalldataDecoder.sol [hunk 5]
```diff
@@ -237,38 +218,74 @@ library CalldataDecoder {
...
+            res.length := length
+            res.offset := offset
         }
         if (_bytes.length < length + relativeOffset) revert SliceOutOfBounds();
     }
 
     /// @notice Decode the `_arg`-th element in `_bytes` as `bytes`
-    /// @param _bytes The input bytes string to extract a bytes string from
+    /// @param _bytes The input bytes string to extract a bytes array from
     /// @param _arg The index of the argument to extract
```
Comment (snreynolds) [inline src/libraries/CalldataDecoder.sol:247]: can you add a comment for when this is expected to revert with the SliceOutOfBounds error?

**Comment by kmbarry1** [comment]
I think `actions` and `params` could overlap even with the improved checks, e.g. if the head of `actions` just points at one of the elements of `params`. I'm not sure that's problematic per se (based on usage in the code, which is read-only AFAICT), but it is weird.

**Review by snreynolds — COMMENTED** [review]
i have reviewed just the tests which look good to me!

**Comment by hensha256** [comment]
closing for #346 due to @kmbarry1 's comment above

**Pull request closed without merging**